### ISSUE TITLE
Soliciting feedback on plans to merge Picard + GATK4

### ISSUE THREAD

**Issue by vdauwera** [broadinstitute/picard#647]
For many years now we’ve been hearing from users of both GATK and Picard about how they’d love to see the two projects **unite into a single "toolkit-to-rule-them-all"**, for the sake of **user convenience**, to promote **consistency across tools**, and to **minimize duplication of effort**. 

With the advent of **GATK 4** this suddenly became **a real possibility**, as the decision was made to start the new GATK codebase from the Picard base classes rather than the old GATK 3.x base classes. This allows for **free-form Picard-style tools and GATK “walkers”** built upon an engine traversal to **peacefully co-exist** within the same framework. Last year, a Picard engineer successfully **ported all Picard tools to the GATK 4 codebase** with only minor changes to the tools themselves. More recently, efforts have been made to **harmonize the build systems** of the two projects, resulting in Picard’s recent move to gradle. 

Importantly, the **core GATK 4 codebase** at https://github.com/broadinstitute/gatk is released entirely under the **BSD 3-clause license**, a big improvement over the confusing licensing situation in GATK 3.x, with its mix of open-source and proprietary licenses within the same repository -- and that is where any Picard tools moved to the GATK 4 codebase would live, **remaining fully open-sourced and free for all**.

As **all of the technical pieces are now in place** to allow for a merger of the two projects (with the guarantee that the open-source nature of Picard code will be preserved) we are **soliciting feedback from the Picard developer community** about the prospect of a union with GATK. Would people here be generally in favor of such a move? Are there any strong objections to this idea? Any concerns that should be addressed _before_ we head any further down this path?

To help start the discussion, **here is the case in favor of a GATK4 + Picard merger** as we see it:
- It would greatly **cut down on duplication of effort** and re-invention of functionality already existing in one project or the other -- Picard devs could directly use open-source GATK functionality/utilities, and vice versa.
- **Bug fixes and new features** such as CRAM support could **more easily propagate** to all the tools in the GATK+Picard ecosystem, and the tools overall could be made **more consistent and easier to use**.
- Picard and GATK devs could better **share knowledge and give each other feedback**, and with more eyes looking at the code and more shared developer resources available the **overall quality of the tools should increase**.
- **End-users’ lives** would be greatly simplified by having **all of the tools in one place**, with a unified interface, a common set of conventions and a single place to go to for documentation, help and updates. 
- Along similar lines, the ability of the GATK/Picard support team to **provide in-depth support for the tools** would be greatly enhanced. Among other things this would speed up plans to provide **developer support and API documentation**. 
- Only **minimal changes to the Picard tools** would be required. The biggest change would be a switch to standard POSIX-style command-line arguments (eg., `--argName value` instead of `argName=value`). 
- The **work of porting has already largely been done**, and in the process, test coverage of the ported versions of the Picard tools has been improved and bugs have been fixed.

In short: we’re stronger together than apart, right? :)

We realize that such a move is never trivial, but we're prepared to put resources into making the migration as painless as possible for everyone, and we believe that it would take us all to a better situation than what we have right now. 

What do you think? Let the discussion begin!

**Comment by lordzappo** [comment]
![bitmoji](https://render.bitstrips.com/v2/cpanel/9946850-132012173_2-s1-v1.png?transparent=1&width=246)

**Comment by tfenne** [comment]
I'm somewhat ambivalent about this.  I think there are clear benefits as outlined by @vdauwera  but there are also downsides.  The biggest of which I see as:

1. The existing Picard codebase is likely to become unmaintained, stranding users who can't/don't want to take the time to update all their pipelines and scripts to use the new GATK/Picard combo.  I say this because of the whole change to Posix style arguments.  I think if there were a way to enable both styles of invocation for some time, this would make the transition easier.  In my own case I have a ton of pipelines and scripts that use both Picard and GATK.  I've never found it particularly onerous to have two JARs and two styles of invocation (my pipelines tend to invoke many other tools to, so invoking 11 vs 12 different tools is not really different).  I would find it onerous to have to convert everything in order to move to the next version of Picard, without some way to do it piecemeal. 

2. I haven't looked in a while but the last time I did the GATK 4 build process was much more time consuming than the Picard one because of the volume of code, but moreso the volume of dependencies.  The two projects have different attitudes toward dependencies (Picard tries to minimize them, GATK seems to use half of maven central).  As a developer and user of Picard I don't want to have to wait longer for my builds, and I don't want a final JAR that is many times larger.

3. Despite the fact that the core of GATK is open there is still a _ton_ of confusion outside of the walls of the Broad about licensing.  I can guarantee that some non-trivial fraction of Picard users will hear about this indirectly and _assume_ that this means they have to pay for Picard now too.

As I said up top, I'm ambivalent; I see the advantages, but also some disadvantages.  I would like to see a healthy discussion here before trying to drive to a conclusion.

**Comment by droazen** [comment]
1. I think we'd be open to the idea of maintaining both styles of argument syntax for a transitional period to ease the pain for script/pipeline writers.

2. GATK4 builds much faster than GATK3, and the difference between building Picard alone vs. building GATK4 + Picard is not that great:

Time to build GATK4 (**including** ported Picard tools) from a clean checkout: 22.314s  
Time to build Picard alone from a clean checkout: 13.151 secs

(This is with dependencies cached locally after an initial build -- the very first build of GATK4 does take longer, naturally)

3. I'll let @vdauwera respond to this one, but I think it's a communication issue that could be solved with the help of her support team (which is already handling both GATK and Picard support).

**Comment by yfarjoun** [comment]
Thanks for this post. I see that is has already sparked a healthy discussion. Hopefully it will bring about an understanding and (eventually) an agreement.

I think some more things should be clarified (some overlap with what @tfenne wrote, but I wrote my draft last night...so I'll risk repeating some of his points):

1. picard is currently used (also) as the basis of other projects that use its public classes (not just CLPs). How difficult will it be for such a project to switch over to being based on GATK4?
2. Traditionally, GATK was a large distribution, took long to compile and test, and was difficult to contribute to externally; picard was the opposite. Is GATK4 committing to a picard-style repo/project regarding building and testing?
3. You mentioned the changes in commandline style, what about metrics files? Picard and GATK used differently-styled metrics-files. What will GATK4 use? 
4. As an OSS project, what will be the governance model of GATK4? How will conflicting views be resolved when involving an external developer? Will there be external developers with write access? This is especially important since Picard has lively external collaborations that (I hope) feel like a well-integrated part of the project.
5. The licensing issue can use clarification. Will there be any tools in GATK4 that will have a different license than the BSD one? You mentioned "core GATK 4 codebase" are there tools that are not "core"? It might help if there was an official statement from the Broad regarding this issue.

Regarding  @tfenne's comments on commandline style: If gatk4 implements both styles, why/how would we ever take out the non-posix version? It seems that we will be kicking the can down the road....

**Comment by vdauwera** [comment]
OK, so far we've heard from people on the following concerns; let me know if I'm leaving out anything:

### Migration pain

The idea of maintaining two argument parsing styles is controversial -- we'd love to get more opinions on this. 

For end-users, we're going to be going through some migration pain anyway to help GATK users transition to GATK4. One thing that should help is that we'll be providing lots of example/working pipelines. We're investing heavily in support on this front. 

For developers, I think the best thing we can do is produce developer docs, starting with a detailed migration guide. 

### Build time

See @droazen's comment above -- building GATK4 is akin to building Picard. We've learned a lot from years of pain with classic GATK, and we're not going back. 

### Licensing / confusion around OSS portion of GATK / GATK4

Definitely something we can address with better communication. We haven't talked much about this so far because it was still taking shape, but now a clear picture is emerging. I'll try to write up a full description in the near future that can serve as "official Broad statement" (with lawyer backing!). In the meantime, hopefully this will help: 

The open-source core of GATK4, which I'll call GATK-core from here onwards, encompasses the engine, all the boilerplate, infrastructure, public APIs etc that an external developer would want to develop against, plus Picard tools that have been ported, plus many GATK tools that have been open-sourced (such as the BQSR and VQSR tools). That's the entire codebase you can find at https://github.com/broadinstitute/gatk. In addition to that, there is a separate codebase at https://github.com/broadinstitute/gatk-protected that contains proprietary code for select tools such as HaplotypeCaller and the new CNV tools. 

### Governance model of GATK4-core

Agreed, this is something we should figure out. Would anyone like to suggest a particular model, or namedrop a project that we should look to for inspiration?

### Specific changes like metrics formats

Established Picard metrics formats will continue to be supported. Picard outshines GATK on metrics collection, and wherever there is overlap, I expect we'll go with the Picard equivalent.

**Comment by droazen** [comment]
@yfarjoun To supplement @vdauwera's response above:

1. If the project is pulling down Picard via maven, then it should hopefully not be much more work than changing the dependency from Picard to GATK in the project's build file, and updating imports to new package names. If the latter is a big concern, we could leave the existing package names intact during the transition period. The Picard utilities and tools themselves would be almost unchanged, so there would be no nightmarish re-adjustment to new APIs.

2. We are VERY determined to keep build/test times down in GATK4, after our experience with GATK3. As mentioned above, the difference between `gradle installDist` in GATK4+Picard vs. Picard alone is in seconds, not minutes. Building a full GATK4 jar for distribution does take longer (about 1-2 minutes), but you typically don't need to do that repeatedly during development.

For tests, we have been making a systematic effort to reduce test suite runtime relative to GATK3 by shrinking test inputs, intervals, etc. -- it's currently ~30 minutes instead of several hours. Many of the more expensive tests (eg., `HaplotypeCaller` tests) will live in gatk-protected, rather than gatk, and so will not affect Picard development. We've also purchased a travis CI account that allows us to run 25 concurrent builds, which gives us the option of speeding up the test suite further via increased parallelism. Our target goal is to get overall runtime on travis down to **10 minutes per build**. We could get closer to this goal by running GATK and Picard tests in separate parallel travis builds (they're currently run together).

**Comment by lbergelson** [comment]
@tfenne about your second point.   GATK has a lot more dependencies.  Most of these are to support cloud-stuff, spark and google dependencies that pull in huge apache / google stacks.  This makes the first build time take much longer than picard as dependencies are cached, a few minutes on a reasonably fast internet connection.  Practically, I don't think that this has been a pain point for anyone of the developers.  Things that it might adversely effect, like CI services, typically support dependency caching.  

The build time for gatk with cached dependencies is somewhat longer than picard as @droazen already noted.  It's still pretty fast though, and it should be noted that it already includes building many of the existing picard tools, so it's unlikely to increase much with a merger.  Practically it feels somewhat slower than building picard alone, but it still has a pretty good cycle time.  There are a number of new gradle features that we could likely use to reduce the cycle time if it became to onerous, like modularization of various components to prevent rebuilding and making use of continuous builds.  

The biggest issue that I see is the jar size.  GATK jar's are much larger than picard jars.  Currently in the ~200mb range.  One reason is that we have more dependencies. The second is that we are packaging a number of statically linked native binaries for multiple platforms.  This allows us to distribute make use of optimized C code for some tasks without the users having to worry about installing libraries or setting LD_PATH, but it comes at a cost to file size.  There are things that we could do to minimize the file size, but they're somewhat involved and would take an investment of time or come at the cost of more complicated distribution.  Is file size a significant issue for you?  I don't think it's insurmountable to reduce the gatk distribution size, but it would be some effort, and it will always be larger than Picard alone.  

@yfarjoun  I'm not sure exactly what you mean by a picard-style repo, but GATK4 is a proper public github project.  It's open to forks and pull requests just like Picard is.  It's had 35 contributors so far, mostly from Broad and various Broad collaborators, but we are open to pull requests from the outside world.  Tests run on travis, so they run on pull request from forks as well (except for a few tests that require private cloud keys), and the results are publicly viewable.  It's licensed with BSD-3 clause which is an extremely permissive license similar to Picard's MIT license.  

I think it would be useful if we had at least a partial list of active projects that depend on picard.  It would make it easier to understand what changes are safe to make and what would cause a lot of people problems.

**Comment by droazen** [comment]
For the jar size issue, we could package the native libraries in a separate jar, and slim down the main jar using dependency analysis (only package third-party classes that are actually used). GATK3 does this, which is why its jar is in the 10-15 MB range despite having a huge number of dependencies.

**Comment by nh13** [comment]
Thanks Geraldine for asking the community!

tl;dr: I prefer to keep them separate.

The following influences my preference, and I reserve the right to change my mind, as any good scientist would do, as well as be the contrarian as I have always wanted to be:

1. All of the GATK?
  - It would go a long way to release `HaplotypeCaller` and `GenotypeGvcfs` under BSD , as this would provide a lot of goodwill to folks using GATK.  It would be worth it to me without further discussion if this was the case.
  - It is still not clear that GATK-core (or GATK-public) does not include the two aforementioned tools.  It would also be useful to explain why this is still the case.

2. Governance
  - Who are the folks with the final say on direction and review for GATK+Picard? 
  - How do we reconcile the current different in philosophies about dependencies and other issues we have yet to think about?
  - Is this an open source project owned by the Broad Institute, or the community?
  - It is not clear who makes the final call on combining Picard+GATK?  It is not clear if the Picard community opposes or is ambivalent to this change if this will stop the merger.  Could you provide some transparency on the decision-making process?

3. Brand and Reputation
  - The merger would lose the Picard brand and reputation and tie it with GATK.  I certainly have pride in Picard when I contribute, even for all its warts, and that pride may not be the same when we move to GATK.  I have much less history and ownership over the GATK codebase, especially in terms of some of the legacy decisions about the framework and commercial license for certain tools.  This would reduce my motivation to contribute.

4. Option Parsing Changes
  - There are so many tools and pipelines written to call Picard and would cost many years of developer time across many organizations.  This should not be underestimated.  Perhaps the cost of the move could be mitigated by open sourcing certain tools?

5. Dependencies, Build times, and Testing
  - While minor, the build time difference will slow development down.  Keeping two separate projects makes sense here. 
  - Docker pt1: many dependencies means we need to have two docker commands, one to "cache" the dependencies in one layer and other to check out the branch/commit/tag of choice, update, and build.  This is true when I don't use any of the new GATK functionality (which most of the time I don't). 
  - Docker pt2: long build times means building a docker image is sssssllllloooowwwww.
  - How do outside developers (aka the community) test PRs for inclusion?

6. Package re-naming
  - We did this once with `htsjdk` and `Picard`, and that was painful for us and others. It will be painful yet again.

7. Picard port
  - I am a bit disappointed that we are just now having the discussion about if it is a good idea when "the work of porting has already largely been done".

8. Miscellaneous
  - I am actually overwhelmed by the idea of "looking at" the GATK codebase given my lack of familiarity, and I am not sure your point about the quality would be increased as I am not sure I am properly motivated to contribute either code or review time for it. This signals that the quality may actually decrease.
  - "all of the tools in one place": I use so many other tools, this will never be the case.  I think we just have a philosophical difference here.  I am more in the camp of small purpose-built tools (UNIX-like) that do not have to be in one jar to rule them all.  But I am usually wrong.
 - on a personal note, I am moving a lot of my major tool development into `Scala` (see https://github.com/fulcrumgenomics/fgbio), as even with Java8 improvements I find my ability to rapidly develop tools is improved in `Scala`.  Therefore, my own contributions my be further reduced independent of this proposed change.
  
Hopefully these points add to the discussion, and likely they can be explained away if so desired, so I will not further debate them here. I will say that personally I see my current level of contributions to Picard being reduced with such a move (unless (1) happens!).  Nonetheless, I will adapt and so I would rather do what's best for the community, which reaches beyond the Picard and GATK developers, support staff, and GATK forum users who may see this discussion.

**Comment by tomwhite** [comment]
Re: Governance model of GATK4-core

Hosting at a software foundation is a good way to spread governance across multiple organizations and people. The one I know the best is the Apache Software Foundation (ASF), and there are many successful projects being run there in a fairly vendor-neutral fashion. It's best known for its infrastructure projects (HTTP Server, Hadoop, Spark, etc), but there are also several examples of more scientific software (in the vein of GATK) being hosted there - for example, Apache Open Climate Workbench (https://climate.apache.org/).

More info:
* http://apache.org/foundation/how-it-works.html
* http://theapacheway.com/

**Comment by vdauwera** [comment]
Hi all, just a quick update on this topic -- following the feedback we've received, as well as some internal discussions of how this move could adversely affect our internal operations systems, we're looking at options for reformulating our strategy in a way that would minimize disruption both internally and in the community at large, and address some of the key concerns that have been brought up here and elsewhere.  We'll try to have a more complete update for you when we cycle back around from our second pass at the drawing board.

**Comment by alecw** [comment]
I've been away so I apologize for being late to this discussion.  I have a few thoughts, some of which are merely echoing what others have written above:

* I have a number of workflows that use the Picard-style CLP, both to invoke Picard tools and our own tools that use the Picard CLP.  Converting all my workflows would be a non-trivial effort.  In addition, there are almost 600 subscribers to the Dropseq Google group.  I'm not sure how many of them are active, but any who are would eventually also need to port their workflows.

* I am concerned that I'm going to run into issues building with all the additional dependencies.  I don't use Gradle.  I don't use Maven.  I use plain old Ant, with the dependencies copied into my project.

* I have been burned before by the enormous entourage of features that has been the GATK in the past.  E.g. I spent a while figuring out that jobs were taking a long time because GATK was calling javax.crypto.Cipher.getInstance() so that it could phone home, and for some reason that process needed to list the temp directory, and the temp directory was getting filled up because of something else, etc., etc.  Picard is lean.  Historically, at least, GATK has not been, and I fear I will spend a lot of time dealing with the unintended consequences of something that someone thought was a good idea.

-Alec

**Comment by yfarjoun** [comment]
This issue seems to be resolved given that GATK is ingesting Picard.jar and the two codebases are sharing a parser. Future concerns should be discussed in a separate ticket.
